### ISSUE TITLE
VSCode doesn't recognize file changes

### ISSUE THREAD

**Issue by eetupulkkinen** [microsoft/vscode#230382]
Type: <b>Bug</b>

I don't know a sure way to reproduce the bug. What happens is that I need to 'reload' my vscode daily, usually multiple times.

Usually it happens after pushing my changes to source control, or changing branches. 

What happens is that I make file changes and save the file, the file name doesn't turn yellow and the changed file doesnt pop up in source control. It is fixed by using 'Developer: Reload Window' function in vscode, but it's annoying to do multiple times in a single session.

VS Code version: Code 1.93.0 (4849ca9bdf9666755eb463db297b69e5385090e3, 2024-09-04T13:02:38.431Z)
OS version: Darwin arm64 23.6.0
Modes:

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|Apple M2 Pro (12 x 2400)|
|GPU Status|2d_canvas: enabled<br>canvas_oop_rasterization: enabled_on<br>direct_rendering_display_compositor: disabled_off_ok<br>gpu_compositing: enabled<br>multiple_raster_threads: enabled_on<br>opengl: enabled_on<br>rasterization: enabled<br>raw_draw: disabled_off_ok<br>skia_graphite: disabled_off<br>video_decode: enabled<br>video_encode: enabled<br>webgl: enabled<br>webgl2: enabled<br>webgpu: enabled<br>webnn: disabled_off|
|Load (avg)|1, 2, 2|
|Memory (System)|32.00GB (0.08GB free)|
|Process Argv|--crash-reporter-id 7addafe9-5ed2-4532-af0c-ae917adae4a6|
|Screen Reader|no|
|VM|0%|
</details><details><summary>Extensions (16)</summary>

Extension|Author (truncated)|Version
---|---|---
vscode-apollo|apo|2.3.3
atlascode|atl|3.0.10
vscode-tailwindcss|bra|0.12.11
javascript-ejs-support|Dig|1.3.3
xml|Dot|2.5.1
gitlens|eam|15.5.1
prettier-vscode|esb|11.0.0
copilot|Git|1.235.0
copilot-chat|Git|0.20.3
vscode-docker|ms-|1.29.3
playwright|ms-|1.1.7
vscode-thunder-client|ran|2.27.0
vscode-yaml|red|1.15.0
ejs-snippets|Tao|0.1.0
pdf|tom|1.2.2
pretty-ts-errors|Yoa|0.6.0

</details><details>
<summary>A/B Experiments</summary>

```
vsliv368cf:30146710
vspor879:30202332
vspor708:30202333
vspor363:30204092
vscod805cf:30301675
binariesv615:30325510
vsaa593cf:30376535
py29gd2263:31024239
c4g48928:30535728
azure-dev_surveyone:30548225
2i9eh265:30646982
962ge761:30959799
pythongtdpath:30769146
welcomedialog:30910333
pythonnoceb:30805159
asynctok:30898717
pythonmypyd1:30879173
2e7ec940:31000449
pythontbext0:30879054
accentitlementst:30995554
dsvsc016:30899300
dsvsc017:30899301
dsvsc018:30899302
cppperfnew:31000557
dsvsc020:30976470
pythonait:31006305
dsvsc021:30996838
bdiig495:31013172
a69g1124:31058053
dvdeprecation:31068756
dwnewjupyter:31046869
newcmakeconfigv2:31071590
impr_priority:31102340
nativerepl2:31139839
refactort:31108082
pythonrstrctxt:31112756
flighttreat:31134774
wkspc-onlycs-t:31132770
nativeloc2:31134642
wkspc-ranged-t:31125599
cf971741:31144450
defaultse:31146405
notype1:31150297
5fd0e150:31146321

```

</details>

**Comment by unional** [comment]
Notice the same issue. Need to manually click on the source control refresh button.
![image](https://github.com/user-attachments/assets/0aaff7fe-6068-4953-9c0b-35df6b1b9bdb)

**Comment by xy-23** [comment]
same issue

```
Version: 1.94.0
Commit: d78a74bcdfad14d5d3b1b782f87255d802b57511
Date: 2024-10-02T13:08:12.626Z
Electron: 30.5.1
ElectronBuildId: 10262041
Chromium: 124.0.6367.243
Node.js: 20.16.0
V8: 12.4.254.20-electron.0
OS: Darwin arm64 23.6.0
```

**Comment by lszomoru** [comment]
Please use the steps documented on [the wiki](https://github.com/microsoft/vscode/wiki/File-Watcher-Issues#logging-local) to collect file watcher logs.

**Comment by unional** [comment]
I followed the steps and there are no "File Watcher" trace log.

Here are some of the logs without the filter:

```sh
https://marketplace.visualstudio.com/_apis/public/gallery/extensionquery - begin POST n6i {b: {…}, a: {…}}a: {X-Market-Client-Id: 'VSCode 1.94.2', User-Agent: 'VSCode 1.94.2 (Code)', X-Market-User-Id: 'c46cba86-6efa-4c11-88b2-c0b41ac30f8b', VSCode-SessionId: '4651b5b6747184c219557e87cb2ba4b7a683b5a0ca829a21d91f2737700ff9b6', Content-Type: 'application/json', …}b: {X-Market-Client-Id: 'VSCode 1.94.2', User-Agent: 'VSCode 1.94.2 (Code)', X-Market-User-Id: 'c46cba86-6efa-4c11-88b2-c0b41ac30f8b', VSCode-SessionId: '4651b5b6747184c219557e87cb2ba4b7a683b5a0ca829a21d91f2737700ff9b6', Content-Type: 'application/json', …}[[Prototype]]: Object
log.ts:399 TRACE [network] #26: https://marketplace.visualstudio.com/_apis/public/gallery/extensionquery - end POST 200 {activityid: '57a5e2ae-7815-4994-b4fc-ed6adc469eab', cache-control: 'no-cache, no-store, must-revalidate', content-type: 'application/json; charset=utf-8; api-version=3.0-preview.1', expires: '-1', pragma: 'no-cache', …}
log.ts:399 TRACE MainThreadTreeViews#$refresh currentBranchInfo undefined
log.ts:399 TRACE KeybindingService#dispatch ↓ [ Will dispatch command quickInput.next ]
log.ts:399 TRACE CommandService#executeCommand quickInput.next
log.ts:399 TRACE KeybindingService#dispatch ↓ [ Will dispatch command quickInput.next ]
log.ts:399 TRACE CommandService#executeCommand quickInput.next
log.ts:399 TRACE writeTextAreaState(reason: focusgain)
log.ts:399 TRACE CommandService#executeCommand git.pushForce
log.ts:399 TRACE onWillActivateByEvent:  onCommand:git.pushForce
log.ts:399 TRACE CommandService#executeCommand _setContext
log.ts:409 DEBUG Comments: URIs of continue on comments to add to storage .
2log.ts:399 TRACE CommandService#executeCommand _setContext
log.ts:399 TRACE MainThreadFileSystemEventService#$unwatch(): request to stop watching (session: 0.6570562027914344)
log.ts:399 TRACE MainThreadFileSystemEventService#$watch(): request to start watching uncorrelated (extension: vscode.git, path: file:///Users/hwong/pan/panwds/.git/refs/remotes/origin/r18, recursive: false, session: 0.26773984001931006, excludes: [], includes: ["**/.git/objects/**/**","**/.git/subtree-cache/**/**","**/.hg/store/**/**","**/.ammonite/**","**/.bloop/**","**/.metals/**"])
6

...

log.ts:399 TRACE [network] #26: https://marketplace.visualstudio.com/_apis/public/gallery/extensionquery - end POST 200 {activityid: '57a5e2ae-7815-4994-b4fc-ed6adc469eab', cache-control: 'no-cache, no-store, must-revalidate', content-type: 'application/json; charset=utf-8; api-version=3.0-preview.1', expires: '-1', pragma: 'no-cache', …}
log.ts:399 TRACE MainThreadTreeViews#$refresh currentBranchInfo undefined
log.ts:399 TRACE KeybindingService#dispatch ↓ [ Will dispatch command quickInput.next ]
log.ts:399 TRACE CommandService#executeCommand quickInput.next
log.ts:399 TRACE KeybindingService#dispatch ↓ [ Will dispatch command quickInput.next ]
log.ts:399 TRACE CommandService#executeCommand quickInput.next
log.ts:399 TRACE writeTextAreaState(reason: focusgain)
log.ts:399 TRACE CommandService#executeCommand git.pushForce
log.ts:399 TRACE onWillActivateByEvent:  onCommand:git.pushForce
log.ts:399 TRACE CommandService#executeCommand _setContext
log.ts:409 DEBUG Comments: URIs of continue on comments to add to storage .
2log.ts:399 TRACE CommandService#executeCommand _setContext
log.ts:399 TRACE MainThreadFileSystemEventService#$unwatch(): request to stop watching (session: 0.6570562027914344)
log.ts:399 TRACE MainThreadFileSystemEventService#$watch(): request to start watching uncorrelated (extension: vscode.git, path: file:///Users/hwong/pan/panwds/.git/refs/remotes/origin/r18, recursive: false, session: 0.26773984001931006, excludes: [], includes: ["**/.git/objects/**/**","**/.git/subtree-cache/**/**","**/.hg/store/**/**","**/.ammonite/**","**/.bloop/**","**/.metals/**"])
6```

**Comment by viktorrenkema** [comment]
I had this too, and found out that it was caused by the same issue described here: https://github.com/microsoft/vscode/issues/201099#issuecomment-1909545629.

tldr, branch name is too long. try these settings or better, rename the branch:
`"scm.showIncomingChanges": false`
`"scm.showOutgoingChange": false,`

**Comment by unional** [comment]
Um...for me, it is not related to branch name. I work on `main`.

**Comment by eetupulkkinen** [comment]
> Please use the steps documented on [the wiki](https://github.com/microsoft/vscode/wiki/File-Watcher-Issues#logging-local) to collect file watcher logs.

I've been trying to cause the bug so I could get you some trace logs, but this bug hasn't occured to me in a week, which is weird.
I'll get back to this issue as soon as the bug occurs again. But 'nice' to see that I am not the only one with this bug. 

Thanks for the replies.


> I had this too, and found out that it was caused by the same issue described here: [#201099 (comment)](https://github.com/microsoft/vscode/issues/201099#issuecomment-1909545629).
> 
> tldr, branch name is too long. try these settings or better, rename the branch: `"scm.showIncomingChanges": false` `"scm.showOutgoingChange": false,`

My branch names are not too long either, since it's something like xxx-yyyy.

**Comment by eetupulkkinen** [comment]
Now I got this bug to occur. 

As mentioned before by @unional , no logs showing with 'File Watcher' filter on. Removed the filter.
I tried saving a file that wouldn't turn yellow, and got bunch of logs, but this is the most interesting one

`TRACE [text file model] handleSaveSuccess(102) - setting dirty to false because versionId did not change `

I also have one bugged file that shows yellow (as changed) but it doesnt add my new changes to vscode version control. Tried saving the file and it said the exactly same log.

`TRACE [text file model] handleSaveSuccess(646) - setting dirty to false because versionId did not change `

**Comment by thernstig** [comment]
I am having this issue as well, but I can see the file changed in the Explorer view. It is yellow and shows the number of changes in the file.

The problem is that it does not immediately show up in the **Source Control view**. It does show up if I click refresh.

I am using WSL, so I wonder if it has something to do with that?

@lszomoru please note that the file watching issues wiki seems to contain a bug, so it is hard to know which output channel to view. See https://github.com/microsoft/vscode/issues/232715.

**Comment by thernstig** [comment]
There is a definitely a bug here. I am using Windows 11 with WSL2 enabled. After my computer comes back from hibernation, any changes made does not automatically show up in the **Source Control** view. I have to press refresh for it to show up.

There is no problem visible in the `Server` output channel after having enabled trace logs.

```log
2024-11-04 13:03:31.893 [trace] [File Watcher ('parcel')] [CHANGED] /home/tobias/code/c-test/src/main.c
2024-11-04 13:03:31.943 [trace] [File Watcher ('parcel')] [CHANGED] /home/tobias/code/c-test/src/main.c
2024-11-04 13:03:31.969 [trace] [File Watcher ('parcel')]  >> normalized [CHANGED] /home/tobias/code/c-test/src/main.c
2024-11-04 13:03:46.201 [trace] [File Watcher ('parcel')] [CHANGED] /home/tobias/code/c-test/src/main.c
2024-11-04 13:03:46.251 [trace] [File Watcher ('parcel')] [CHANGED] /home/tobias/code/c-test/src/main.c
2024-11-04 13:03:46.277 [trace] [File Watcher ('parcel')]  >> normalized [CHANGED] /home/tobias/code/c-test/src/main.c
2024-11-04 13:04:07.470 [trace] Started scanning user extensions {"$mid":1,"external":"file:///home/tobias/.vscode-server/data/User/profiles/-2da0915b/extensions.json","path":"/home/tobias/.vscode-server/data/User/profiles/-2da0915b/extensions.json","scheme":"file"}
2024-11-04 13:04:07.517 [trace] Scanned user extensions: 24
```

@lszomoru what else can I do to help here, I am willing to do more extensive troubleshooting. This is detrimental almost to the point of non-workable working with VS Code for WSL remotes when this happens.

**Comment by thernstig** [comment]
@lszomoru this is quite a severe bug/problem working under WSL, could you please look into this?

**Comment by lszomoru** [comment]
@thernstig, sorry for not getting back to you on this until now. Reading through the comments and the logs it seems to me that right now you have a consistent way to reproduce this issue. Could you please confirm that you are able to repro every time the machine comes out of hibernation? Looking at the logs it seems that when you are reproducing the problem you are storing the workspace file in the WSL file system (ex: /home/tobias...) rather than the Windows file system. Could you please also confirm that. Would you be able to share your power management settings to see if I am able to reproduce this on my windows machine along with the exact WSL distro and version that you are using? 

Let's keep the verbose logs on, and also enable verbose logs for the `vscode.git` extension: 
1. Open the Output panel (View -> Output)
2. Switch to the "Git" channel in the dropdown
3. Click on the gear action next to the dropdown, and choose "Trace"
4. Click on the gear action again, and choose "Set as Default"

Reproduce the issue and share the file watcher logs and the contents of the git output channel. Thanks!

**Comment by thernstig** [comment]
@lszomoru I also apologize for not coming back, due to reasons. I plan to look into this soonish again, with everything you ask for. And to try to make sure I can reproduce it 100%. Not sure I can, but will try.
